README.md
# ccc [![Build Status](https://travis-ci.org/andrepolischuk/ccc.svg?branch=master)](https://travis-ci.org/andrepolischuk/ccc)

  Colors converter

## Instalation

  Component(1):

```sh
$ component install andrepolischuk/ccc
```

  Npm:

```sh
$ npm install ccc
```

  Umd:

```html
<script src="https://cdn.rawgit.com/andrepolischuk/ccc/0.4.0/ccc.js"></script>
```

## API

### ccc(str)

  Return [Color](#color) via string

```js
var color = ccc('black')
var color = ccc('#ffffff')
var color = ccc('rgb(255, 0, 0)')
var color = ccc('rgba(255, 0, 0, 0.5)')
var color = ccc('cmyk(100%, 0%, 100%, 0%)')
var color = ccc('hsl(240, 100%, 50%)')
var color = ccc('hsla(240, 100%, 50%, 0.7)')
var color = ccc('hsv(60, 100%, 100%)')
```

### ccc(obj)

  Return [Color](#color) via object

```js
var color = ccc({r: 255, g: 0, b: 0})
var color = ccc({r: 255, g: 0, b: 0, a: 0.5})
var color = ccc({c: 100, m: 0, y: 100, k: 0})
var color = ccc({h: 240, s: 100, l: 50})
var color = ccc({h: 60, s: 100, v: 100})
```

### ccc.rgb(...arguments)

  Return [Color](#color) via arguments

```js
var color = ccc.keyword('black')
var color = ccc.hex('ffffff')
var color = ccc.rgb(255, 0, 0)
var color = ccc.rgb(255, 0, 0, 0.5)
var color = ccc.cmyk(100, 0, 100, 0)
var color = ccc.hsl(240, 100, 50)
var color = ccc.hsl(240, 100, 50, 0.7)
var color = ccc.hsv(60, 100, 100)
```

### ccc.rgb(arr)

  Return [Color](#color) via array

```js
var color = ccc.rgb([255, 0, 0])
var color = ccc.rgb([255, 0, 0, 0.5])
var color = ccc.cmyk([100, 0, 100, 0])
var color = ccc.hsl([240, 100, 50])
var color = ccc.hsl([240, 100, 50, 0.7])
var color = ccc.hsv([60, 100, 100])
```

### Color

#### Color#rgb()

  Return converted color object

```js
color.rgb()
// {r: 255, g: 0, b: 0}

color.cmyk()
// {c: 100, m: 0, y: 100, k: 0}

color.hsl()
// {h: 240, s: 100, l: 50}

color.hsv()
// {h: 60, s: 100, v: 100}
```

#### Color#rgbArray()

  Return converted color array

```js
color.rgbArray()
// [255, 0, 0]

color.cmykArray()
// [100, 0, 100, 0]

color.hslArray()
// [240, 100, 50]

color.hsvArray()
// [60, 100, 100]
```

#### Color#rgbString()

  Return converted color string

```js
color.keyword()
// 'black'

color.hex()
// '#ffffff'

color.rgbString()
// rgb(255, 0, 0)

color.cmykString()
// cmyk(100%, 0%, 100%, 0%)

color.hslString()
// hsl(240, 100%, 50%)

color.hsvString()
// hsv(60, 100%, 100%)
```

#### Color#invert()

  Invert color

```js
ccc('#ffffff').invert().hex()
// #000000
```

#### Color#grayscale()

  Convert color to grayscale

```js
ccc('#669900').grayscale().hex();
// #787878
```

#### Color#average(Color2)

  Calculate average

```js
ccc('#333333').average(ccc('#999999')).hex()
// #666666
```

#### Color#light()

  Check if color is light

```js
ccc('white').light()
// true
```

#### Color#dark()

  Check if color is dark

```js
ccc('black').dark()
// true
```

#### Color#lighten(val)

  Lighten color by `val`

```js
ccc('#000000').lighten(50).hex()
// #080808
```

#### Color#darken(val)

  Darken color by `val`

```js
ccc('#ffffff').darken(50).hex()
// #080808
```

#### Color#saturate(val)

  Saturate color by `val`

```js
ccc('gray').saturate(50).saturation()
// 50
```

#### Color#desaturate(val)

  Desaturate color by `val`

```js
ccc('red').desaturate(50).saturation()
// 50
```

#### Color#red([val])

  Return or set red component of RGB

```js
ccc('red').red()
// 255

ccc('black').red(255).keyword()
// red
```

#### Color#green([val])

  Return or set green component of RGB

```js
ccc('green').green()
// 255

ccc('black').green(255).keyword()
// lime
```

#### Color#blue([val])

  Return or set blue component of RGB

```js
ccc('blue').blue()
// 255

ccc('black').blue(255).keyword()
// blue
```

#### Color#alpha([val])

  Return or set alpha component

```js
ccc('rgba(0, 0, 0, 0.5)').alpha()
// 0.5

ccc('black').alpha(0.5).rgbString()
// rgba(0, 0, 0, 0.5)
```

#### Color#cyan([val])

  Return or set cyan component of CMYK

```js
ccc('aqua').cyan()
// 100

ccc('white').cyan(100).keyword()
// aqua
```

#### Color#magenta([val])

  Return or set magenta component of CMYK

```js
ccc('fuchsia').magenta()
// 100

ccc('white').magenta(100).keyword()
// fuchsia
```

#### Color#yellow([val])

  Return or set yellow component of CMYK

```js
ccc('yellow').cyan()
// 100

ccc('white').cyan(100).keyword()
// yellow
```

#### Color#key([val])

  Return or set key component of CMYK

```js
ccc('black').key()
// 100

ccc('white').key(100).keyword()
// black
```

#### Color#hue([val])

  Return or set hue component of HSL

```js
ccc('red').hue()
// 0

ccc('lime').hue(240).keyword()
// blue
```

#### Color#saturation([val])

  Return or set saturation component of HSL

```js
ccc('lime').saturation()
// 100

ccc('gray').saturation(100).keyword()
// red
```

#### Color#lightness([val])

  Return or set lightness component of HSL

```js
ccc('blue').lightness()
// 50

ccc('lime').lightness(25).keyword()
// green
```

#### Color#value([val])

  Return or set value component of HSB

```js
ccc('yellow').value()
// 100

ccc('lime').value(50).keyword()
// green
```

## License

  MIT


HISTORY.md
# 0.4.0 / 2015.04.16

  * Add travis
  * Add mnml-tpl instead of fmt
  * Add local deps instead of global
  * Add get and set color components
  * Add transformation functions
  * Rename Color#key to Color#keyword
  * Remove array and object for hex and keyword output
  * Fix browserify support
  * Refactor

# 0.3.0 / 2015.03.18

  * Add parsing from object
  * Add alpha to Color
  * Add output to object and css string
  * Add rgba, hsl(a), hsv(a) models support, fix #2
  * Add parsing from array
  * Add parsing color name, fix #4
  * Add output to array
  * Fix #3 parsing from css color string
  * Fix parsing hex
  * Fix convert to color name
  * Fix hsl and hsv conversions
  * Fix negative deg values
  * Fix fractional values
  * Extract fmt
  * Remove string parsing via ccc methods
  * Update test

# 0.2.0 / 2015.02.19

  * Add average calculation #1


index.js

'use strict';

/**
 * Module dependencies
 */

var each = require('ea');
var tpl = require('mnml-tpl');

try {
  var type = require('type');
} catch (err) {
  var type = require('component-type');
}

var conversions = require('./lib/conversions');

/**
 * Color matchers
 */

var matchers = {};
matchers.keyword = /^([A-Za-z]+)$/;
matchers.hex = /^#([a-f0-9]{3}([a-f0-9]{3})?)$/i;
matchers.rgb = /^rgba?\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,?\s*([\d\.]+)?\s*\)$/;
matchers.cmyk = /^cmyk\(\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*\)$/;
matchers.hsl = /^hsla?\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,?\s*([\d\.]+)?\s*\)?$/;
matchers.hsv = /^hsva?\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*,?\s*([\d\.]+)?\s*\)?$/;

/**
 * Color format
 */

var format = {};
format.keyword = [':k'];
format.hex = ['#:k'];
format.rgb = ['rgb(:r, :g, :b)', 'rgba(:r, :g, :b, :a)'];
format.cmyk = ['cmyk(:c%, :m%, :y%, :k%)'];
format.hsl = ['hsl(:h, :s%, :l%)', 'hsla(:h, :s%, :l%, :a)'];
format.hsv = ['hsv(:h, :s%, :v%)', 'hsva(:h, :s%, :v%, :a)'];

/**
 * Expose parser
 *
 * @param  {String} color
 * @return {Object}
 * @api public
 */

module.exports = function(color) {
  var model;

  each(matchers, function(matcher, m) {
    var key = /(hex|keyword)/.test(m) ? m : m.charAt(2);
    if (type(color) === 'string' && matcher.test(color)) model = m;
    if (type(color) === 'object' && key in color) model = m;
  });

  if (!model) return;

  var matcher = matchers[model];
  if (type(color) === 'object') return parse(model, toArray(color));
  return parse(model, matcher.exec(color).splice(1));
};

/**
 * Expose parser via model
 *
 * @param  {Array} val
 * @param  {Arguments}
 * @return {Object}
 * @api public
 */

each(matchers, function(matcher, model) {
  module.exports[model] = function(val) {
    if (type(val) === 'array') return parse(model, val);
    return parse.apply(null, [model].concat([].splice.call(arguments, 0)));
  };
});

/**
 * Parse color
 *
 * @param  {String} model
 * @param  {Array} val
 * @param  {Arguments}
 * @return {Object}
 * @api private
 */

function parse(model, val) {
  if (type(val) === 'array') return parse.apply(null, [model].concat(val));
  var args = [].splice.call(arguments, 1);
  return new Color(conversions[model + '2rgb'].apply(null, args));
}

/**
 * Convert array to object via keys
 *
 * @param  {String|Array} keys
 * @param  {Array} arr
 * @return {Object}
 * @api private
 */

function toObject(keys, arr) {
  var obj = {};
  each(keys, function(char, i) {
    if (isDefined(arr[i])) obj[char] = arr[i];
  });
  return obj;
}

/**
 * Convert object to values array
 *
 * @param  {Object} obj
 * @return {Array}
 * @api private
 */

function toArray(obj) {
  var arr = [];
  each(obj, function(val) {
    if (isDefined(val)) arr.push(val);
  });
  return arr;
}

/**
 * Check defining
 *
 * @param  {Mixed} val
 * @return {Boolean}
 * @api private
 */

function isDefined(val) {
  return !/(null|nan|undefined)/.test(type(val));
}

/**
 * Color
 *
 * @param {Array} color
 * @api public
 */

function Color(color) {
  this.vals = toObject('rgba', color);
}

/**
 * Return converted color array
 *
 * @param  {String} model
 * @return {Array}
 * @api public
 */

Color.prototype.getArray = function(model) {
  var arr = conversions['rgb2' + model].apply(null, toArray(this.vals));
  if (arr.length > 3 && !isDefined(arr[3])) arr.pop();
  return arr;
};

/**
 * Return converted color object
 *
 * @param  {String} model
 * @return {Object}
 * @api public
 */

Color.prototype.getObject = function(model) {
  var arr = this.getArray(model);
  var keys = /(keyword|hex)/.test(model) ? 'k' : model;
  if (arr.length > keys.length) keys += 'a';
  return toObject(keys, arr);
};

/**
 * Return converted color string
 *
 * @param  {String} model
 * @return {String}
 * @api public
 */

Color.prototype.getString = function(model) {
  var obj = this.getObject(model);
  var alpha = 'a' in obj;
  var fmt = format[model][alpha ? 1 : 0];
  return tpl(fmt)(obj);
};

/**
 * Return converted to HEX and Keyword
 *
 * @return {String}
 * @api public
 */

each(['keyword', 'hex'], function(model) {
  Color.prototype[model] = function() {
    return this.getString(model);
  };
});

/**
 * Return converted to RGB, CMYK, HSL and HSV
 *
 * @return {Mixed}
 * @api public
 */

each(['rgb', 'cmyk', 'hsl', 'hsv'], function(model) {
  Color.prototype[model] = function() {
    return this.getObject(model);
  };

  Color.prototype[model + 'Array'] = function() {
    return this.getArray(model);
  };

  Color.prototype[model + 'String'] = function() {
    return this.getString(model);
  };
});

/**
 * To grayscale
 *
 * @return {Object}
 * @api public
 */

Color.prototype.grayscale = function() {
  var rgb = this.vals;
  rgb.r = rgb.g = rgb.b = Math.round(
    0.299 * rgb.r +
    0.587 * rgb.g +
    0.114 * rgb.b
  );
  return this;
};

/**
 * Invert color
 *
 * @return {Object}
 * @api public
 */

Color.prototype.invert = function() {
  var rgb = this.vals;
  rgb.r = 255 - rgb.r;
  rgb.g = 255 - rgb.g;
  rgb.b = 255 - rgb.b;
  return this;
};

/**
 * Average color
 *
 * @param  {Object} color
 * @return {Object}
 * @api public
 */

Color.prototype.average = function(color) {
  var rgb = this.vals;
  rgb.r = Math.round((rgb.r + color.vals.r) / 2);
  rgb.g = Math.round((rgb.g + color.vals.g) / 2);
  rgb.b = Math.round((rgb.b + color.vals.b) / 2);
  return this;
};

/**
 * Check light color
 *
 * @return {Boolean}
 * @api public
 */

Color.prototype.light = function() {
  return this.lightness() >= 50;
};

/**
 * Check dark color
 *
 * @return {Boolean}
 * @api public
 */

Color.prototype.dark = function() {
  return this.lightness() < 50;
};

/**
 * Lighten color
 *
 * @param {Number} val
 * @return {Object}
 * @api public
 */

Color.prototype.lighten = function(val) {
  val += this.lightness();
  this.lightness(val > 100 ? 100 : val);
  return this;
};

/**
 * Darken color
 *
 * @param {Number} val
 * @return {Object}
 * @api public
 */

Color.prototype.darken = function(val) {
  val -= this.lightness();
  this.lightness(val > 0 ? 0 : -val);
  return this;
};

/**
 * Saturate color
 *
 * @param {Number} val
 * @return {Object}
 * @api public
 */

Color.prototype.saturate = function(val) {
  val += this.saturation();
  this.saturation(val > 100 ? 100 : val);
  return this;
};

/**
 * Desaturate color
 *
 * @param {Number} val
 * @return {Object}
 * @api public
 */

Color.prototype.desaturate = function(val) {
  val -= this.saturation();
  this.saturation(val > 0 ? 0 : -val);
  return this;
};

/**
 * Return or set RGB components
 *
 * @param  {Number} val
 * @return {Number|Object}
 * @api public
 */

each(['red', 'green', 'blue', 'alpha'], function(prop) {
  Color.prototype[prop] = function(val) {
    return this.component('rgb', prop, val);
  };
});

/**
 * Return or set CMYK components
 *
 * @param  {Number} val
 * @return {Number|Object}
 * @api public
 */

each(['cyan', 'magenta', 'yellow', 'key'], function(prop) {
  Color.prototype[prop] = function(val) {
    return this.component('cmyk', prop, val);
  };
});

/**
 * Return or set HSL components
 *
 * @param  {Number} val
 * @return {Number|Object}
 * @api public
 */

each(['hue', 'saturation', 'lightness'], function(prop) {
  Color.prototype[prop] = function(val) {
    return this.component('hsl', prop, val);
  };
});

/**
 * Return or set HSV components
 *
 * @param  {Number} val
 * @return {Number|Object}
 * @api public
 */

Color.prototype.value = function(val) {
  return this.component('hsv', 'value', val);
};

/**
 * Return or set component
 *
 * @param  {String} model
 * @param  {String} prop
 * @param  {Number} val
 * @return {Number|Object}
 * @api public
 */

Color.prototype.component = function(model, prop, val) {
  if (type(val) !== 'number') return this.get(model, prop);
  return this.set(model, prop, val);
};

/**
 * Return component
 *
 * @param  {String} model
 * @param  {String} prop
 * @return {Number}
 * @api public
 */

Color.prototype.get = function(model, prop) {
  var color = conversions['rgb2' + model].apply(null, toArray(this.vals));
  var index = prop === 'alpha' ? 3 : model.indexOf(prop.charAt(0));
  return color[index];
};

/**
 * Set component
 *
 * @param  {String} model
 * @param  {String} prop
 * @param  {Number} val
 * @return {Object}
 * @api public
 */

Color.prototype.set = function(model, prop, val) {
  var color = conversions['rgb2' + model].apply(null, toArray(this.vals));
  var index = prop === 'alpha' ? 3 : model.indexOf(prop.charAt(0));
  color[index] = val;
  this.vals = toObject('rgba', conversions[model + '2rgb'].apply(null, color));
  return this;
};
